xcube/core/resampling/temporal.py

```diff
@@ -127,43 +154,247 @@ def resample_in_time(dataset: xr.Dataset,
+        if replacement['month'] > 12:
+            replacement['month'] -= 12
+            replacement['year'] += 1
+
+    while replacement['month'] > 12:
+        replacement['month'] -= 12
+        replacement['year'] += 1
+
+    return pd.Timestamp(timestamp.replace(**replacement))
+
+
+def _get_previous_timestamp(timestamp, time_unit, time_value, half) \
+        -> pd.Timestamp:
+    # Retrieves the timestamp preceding the passed timestamp according to the
+    # given time unit and time value.
+    # If half is True, the timestamp halfway between the timestamp and the
+    # previous timestamp (which is not necessarily halfway between the two)
+    # is returned
+    increments = _get_increments(timestamp, time_unit, time_value, half)
+    replacement = dict(
+        year=timestamp.year - increments[0],
+        month=timestamp.month - increments[1],
+        day=timestamp.day - increments[2],
+        hour=timestamp.hour - increments[3]
+    )
+
+    while replacement['hour'] < 0:
+        replacement['hour'] += 24
+        replacement['day'] -= 1
+
+    while replacement['day'] < 1:
+        replacement['month'] -= 1
+        if replacement['month'] < 1:
+            replacement['month'] += 12
+            replacement['year'] -= 1
+        replacement['day'] += _days_of_month(replacement['year'],
+                                            replacement['month'] % 12)
+
+    while replacement['month'] < 1:
+        replacement['month'] += 12
+        replacement['year'] -= 1
+
+    return pd.Timestamp(timestamp.replace(**replacement))
+
+
+def _get_increments(timestamp, time_unit, time_value, half) -> List[int]:
+    # Determines the increments for year, month, day, and hour to be applied
+    # to a timestamp
+    if not half:
+        return _tune_increments(TIMEUNIT_INCREMENTS[time_unit], time_value)
+    if time_value % 2 == 0:
+        time_value /= 2
+        return _tune_increments(TIMEUNIT_INCREMENTS[time_unit],
+                                int(time_value))
+    if time_unit in HALF_TIMEUNIT_INCREMENTS:
+        return _tune_increments(HALF_TIMEUNIT_INCREMENTS[time_unit],
+                                time_value)
+    if time_unit == 'QS':
+        num_months = 3
+    else:
+        num_months = 1
+    import math
+    month = int(math.floor((num_months * time_value) / 2))
+    days = _days_of_month(timestamp.year, month)
+    if days % 2 == 0:
+        hours = 0
+    else:
+        hours = 12
+    days = int(math.floor(days / 2)) - 1
+    return [0, month, days, hours]
+
+
+def _tune_increments(incrementors, time_value):
+    incrementors = [i * time_value for i in incrementors]
+    return incrementors
+
+
+def _days_of_month(year: int, month: int):
+    if month in [1, 3, 5, 7, 8, 10, 12]:
+        return 31
+    if month in [4, 6, 9, 11]:
+        return 30
+    if year % 4 != 0:
+        return 28
+    if year % 400 == 0:
+        return 29
+    if year % 100 == 0:
+        return 28
+    return 28
```

Comment:
Can you pls explain why all this code is needed? Why can't we use pandas/xarray here?

Reply:
> See here http://xarray.pydata.org/en/stable/examples/monthly-means.html

[DatasetResample](https://xarray.pydata.org/en/stable/generated/xarray.core.resample.DatasetResample.html) and [DataArrayResample](https://xarray.pydata.org/en/stable/generated/xarray.core.resample.DataArrayResample.html) both return GroupBy objects, so basically we are already using this. The problem is not the grouping itself but the determination of the exact times of the center of the sampling period and the bounds.
I have changed the code so that these are now determined using xarray's groupby.